pkg/oc/admin/diagnostics/diagnostics/cluster/app_create/main.go

```diff
@@ -0,0 +1,359 @@
+	}
+	if d.project == "" {
+		// generate a project if not specified
+		d.project = names.SimpleNameGenerator.GenerateName(d.projectBase)
+	} else {
+		// when an existing project is specified, deleting it is likely to surprise the user, so don't
+		d.keepProject = true
+	}
+	if errs := kvalidation.IsDNS1123Label(d.project); len(errs) > 0 {
+		return fmt.Errorf("invalid project name '%s' for AppCreate: %v", d.project, errs)
+	}
+	// TODO: also test that route is valid under DNS952
+
+	// app management
+	if d.appName == "" {
+		return fmt.Errorf("%s app name cannot be empty", AppCreateName)
+	}
+	if errs := kvalidation.IsDNS1123Label(d.appName); len(errs) > 0 {
+		return fmt.Errorf("invalid app name '%s' for AppCreate: %v", d.appName, errs)
+	}
+	if err := kvalidation.IsValidPortNum(d.appPort); err != nil {
+		return fmt.Errorf("invalid app port %d for AppCreate: %v", d.appPort, err)
+	}
+	d.label = map[string]string{"app": d.appName}
+	d.labelSelector = fmt.Sprintf("app=%s", d.appName)
+
+	d.skipRouteConnect = d.skipRouteConnect || d.skipRouteTest // don't try to connect to route if skipping route test
+
+	return nil
+}
+
+func (d *AppCreate) CanRun() (bool, error) {
+	if d.SARClient == nil || d.AppsClient == nil || d.KubeClient == nil || d.ProjectClient == nil || d.RoleBindingClient == nil {
+		return false, fmt.Errorf("missing at least one client")
+	}
+	if d.PreventModification {
+		return false, fmt.Errorf("requires modifications: create a project and application")
+	}
+	return util.UserCan(d.SARClient, &authorization.ResourceAttributes{
+		Verb:     "create",
+		Group:    kapi.GroupName,
+		Resource: "namespace",
+	})
+}
+
+func (d *AppCreate) Check() types.DiagnosticResult {
+	d.out = types.NewDiagnosticResult(AppCreateName)
+	done := make(chan bool, 1)
+
+	// Jump straight to clean up if there is an interrupt/terminate signal while running diagnostic
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		<-sig
+		d.out.Warn("DCluAC001", nil, "Received interrupt; aborting diagnostic")
+		done <- true
+	}()
+
+	// The actual diagnostic logic
+	go func() {
+		d.result.BeginTime = jsonTime(time.Now())
+		defer func() {
+			d.result.EndTime = jsonTime(time.Now())
+			d.result.TotalDuration = d.result.EndTime.Sub(d.result.BeginTime)
+			done <- true
+		}()
+		if !d.prepareForApp() || !d.createAndTestAppDC() || !d.createAndTestService() {
+			return // without success
+			// NOTE: even if we won't try to connect, we still create the service to test for endpoints
+		}
+		if d.skipRouteTest {
+			d.out.Debug("DCluAC002", "skipping route creation and testing as requested")
+			// however if we just skip connection testing we still create and test for admission
+		} else {
+			d.createAndTestRoute()
+		}
+		d.result.Success = allSucceeded(&d.result.App, &d.result.Service, &d.result.Route)
+	}()
+
+	<-done // wait until either finishes
+	signal.Stop(sig)
+	d.logResult()
```

Comment:
this can run in parallel with the assignment on line 268-269 if interrupt is received, and crash with a data race error

Reply:
If you hit interrupt, how concerned are you about race conditions crashing diagnostics? I'm not sure it's even possible here - yes, data could conceivably be written into the object while the result is being logged, but would you get anything worse than bad output? - however it doesn't seem like an important edge case. But I'd be happy to use a better pattern for handling interrupts if one is known. I think the rest of the product... probably just exits?

Reply:
I wasn't sure if an individual diagnostic could get interrupted and the overall process was expected to keep going

Reply:
yes, that's the idea... it moves on to run the next diagnostic if any. but at that point nothing is going to use the result from the previous diagnostic. Part of the reason you want it to keep going is to give it a chance to clean up the resources it created, so there's an actual benefit...